glib/src/subclass/types.rs

```diff
@@ -233,23 +234,60 @@ impl TypeData {
         self.parent_class
     }
 
-    /// Returns a pointer to the interface implementation specific data.
+    /// Returns a pointer to the class implementation specific data.
     ///
-    /// This is used for interface implementations to store additional data.
-    pub fn get_interface_data(&self, type_: ffi::GType) -> ffi::gpointer {
+    /// This is used for class implementations to store additional data.
+    pub fn get_class_data<T: Any + Send + Sync + 'static>(&self, type_: ffi::GType) -> Option<&T> {
         unsafe {
-            if self.interface_data.is_null() {
-                return ptr::null_mut();
+            if self.class_data.is_none() {
+                return None;
             }
+            let map = self.class_data.unwrap();
+            map.as_ref().get(&type_).map(|v| v.downcast_ref()).flatten()
+        }
+    }
 
-            for &(t, p) in &(*self.interface_data) {
-                if t == type_ {
-                    return p;
-                }
-            }
+    /// Gets a mutable reference of the class implementation specific data
+    ///
+    /// # Safety
+    ///
+    /// This can only be executed before initializing the type data
+    pub unsafe fn get_class_data_mut<T: Any + Send + Sync>(
+        &mut self,
+        type_: ffi::GType,
+    ) -> Option<&mut T> {
+        self.class_data
+            .as_mut()
+            .map(|hashmap| {
+                hashmap
+                    .as_mut()
+                    .get_mut(&type_)
+                    .map(|v| v.downcast_mut())
+                    .flatten()
+            })
+            .flatten()
+    }
 
-            ptr::null_mut()
+    /// Sets class specific implementation data
+    ///
+    /// # Safety
+    ///
+    /// Note the set_class_data can only used before type initialization
+    ///
+    /// # Panics
+    ///
+    /// If the class_data contains a data for the specified `type_`
+    pub unsafe fn set_class_data<T: Any + Send + Sync>(&mut self, type_: ffi::GType, data: T) {
+        if self.class_data.is_none() {
+            self.class_data = ptr::NonNull::new(Box::into_raw(Box::new(HashMap::new())));
+        }
+        if self.class_data.unwrap().as_ref().get(&type_).is_some() {
+            panic!("The class_data already contains a key for {}", type_);
         }
+        self.class_data
+            .unwrap()
```

Comment:
Unnecessary unwrap here and 3 lines above

Reply:
Am I supposed to map/and_then here as well instead of unwrap?

Reply:
That or you could match or use `?`

Reply:
Just do it consistently in all 3 functions :)

Reply:
Updated to make it a bit nicer, We can't make use of `?` and match doesn't make much sense I guess as we check few lines above that there's something in self.class_data first before attempting to read it's content